// @tensorflow/tfjs-node/dist/node.js
"use strict";
/**
 * @license
 * Copyright 2018 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Public API symbols under the tf.node.* namespace.
 */
var callbacks_1 = require("./callbacks");
var image_1 = require("./image");
var saved_model_1 = require("./saved_model");
var tensorboard_1 = require("./tensorboard");
exports.node = {
    decodeImage: image_1.decodeImage,
    decodeBmp: image_1.decodeBmp,
    decodeGif: image_1.decodeGif,
    decodePng: image_1.decodePng,
    decodeJpeg: image_1.decodeJpeg,
    encodeJpeg: image_1.encodeJpeg,
    encodePng: image_1.encodePng,
    summaryFileWriter: tensorboard_1.summaryFileWriter,
    tensorBoard: callbacks_1.tensorBoard,
    getMetaGraphsFromSavedModel: saved_model_1.getMetaGraphsFromSavedModel,
    getNumOfSavedModels: saved_model_1.getNumOfSavedModels,
    loadSavedModel: saved_model_1.loadSavedModel
};
