// @tensorflow/tfjs-node/dist/kernels/ResizeBilinearGrad.js
"use strict";
/**
 * @license
 * Copyright 2020 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tfjs_1 = require("@tensorflow/tfjs");
var nodejs_kernel_backend_1 = require("../nodejs_kernel_backend");
exports.resizeBilinearGradConfig = {
    kernelName: tfjs_1.ResizeBilinearGrad,
    backendName: 'tensorflow',
    kernelFunc: function (args) {
        var _a = args.inputs, images = _a.images, dy = _a.dy;
        var backend = args.backend;
        var alignCorners = args.attrs.alignCorners;
        var opAttrs = [
            nodejs_kernel_backend_1.createTensorsTypeOpAttr('T', images.dtype), {
                name: 'align_corners',
                type: backend.binding.TF_ATTR_BOOL,
                value: alignCorners
            }
        ];
        return backend.executeSingleOutput(tfjs_1.ResizeBilinearGrad, opAttrs, [dy, images]);
    }
};
